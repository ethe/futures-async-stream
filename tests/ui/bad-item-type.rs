// SPDX-License-Identifier: Apache-2.0 OR MIT

#![feature(generators)]

use futures_async_stream::stream;

#[stream(item = Option<i32>)]
async fn option() {
    let val = Some(42);
    if val.is_none() {
        yield None;
        return;
    }
    let val = val.unwrap();
    yield val; //~ ERROR mismatched types
}

#[stream(item = Option<i32>, boxed)]
async fn option_boxed() {
    let val = Some(42);
    if val.is_none() {
        yield None;
        return;
    }
    let val = val.unwrap();
    yield val; //~ ERROR mismatched types
}

#[stream(item = (i32, i32))]
async fn tuple() {
    if false {
        yield 3;
    }
    yield (1, 2) //~ ERROR mismatched types
}

#[stream(item = (i32, i32), boxed)]
async fn tuple_boxed() {
    if false {
        yield 3;
    }
    yield (1, 2) //~ ERROR mismatched types
}

fn main() {}
