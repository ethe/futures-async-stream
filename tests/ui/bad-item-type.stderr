error[E0308]: mismatched types
  --> $DIR/bad-item-type.rs:13:11
   |
13 |     yield val; //~ ERROR mismatched types
   |           ^^^
   |           |
   |           expected enum `Option`, found integer
   |           help: try using a variant of the expected enum: `futures_async_stream::__private::Some(val)`
   |
   = note: expected enum `Option<_>`
              found type `{integer}`
note: return type inferred to be `Option<_>` here
  --> $DIR/bad-item-type.rs:10:9
   |
10 |         return;
   |         ^^^^^^

error[E0308]: mismatched types
  --> $DIR/bad-item-type.rs:32:11
   |
32 |     yield (1, 2) //~ ERROR mismatched types
   |           ^^^^^^ expected integer, found tuple
   |
   = note: expected type `{integer}`
             found tuple `({integer}, {integer})`

error[E0271]: type mismatch resolving `<impl futures_async_stream::__private::stream::Stream as futures_async_stream::__private::stream::Stream>::Item == (i32, i32)`
  --> $DIR/bad-item-type.rs:27:1
   |
27 | #[stream(item = (i32, i32))]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected tuple, found integer
   |
   = note: expected tuple `(i32, i32)`
               found type `{integer}`
   = note: this error originates in the attribute macro `stream` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
  --> $DIR/bad-item-type.rs:24:11
   |
24 |     yield val; //~ ERROR mismatched types
   |           ^^^
   |           |
   |           expected enum `Option`, found integer
   |           help: try using a variant of the expected enum: `futures_async_stream::__private::Some(val)`
   |
   = note: expected enum `Option<_>`
              found type `{integer}`
note: return type inferred to be `Option<_>` here
  --> $DIR/bad-item-type.rs:21:9
   |
21 |         return;
   |         ^^^^^^

error[E0308]: mismatched types
  --> $DIR/bad-item-type.rs:40:11
   |
40 |     yield (1, 2) //~ ERROR mismatched types
   |           ^^^^^^ expected integer, found tuple
   |
   = note: expected type `{integer}`
             found tuple `({integer}, {integer})`

error[E0271]: type mismatch resolving `<impl futures_async_stream::__private::stream::Stream as futures_async_stream::__private::stream::Stream>::Item == (i32, i32)`
  --> $DIR/bad-item-type.rs:35:1
   |
35 | #[stream(item = (i32, i32), boxed)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected tuple, found integer
   |
   = note: expected tuple `(i32, i32)`
               found type `{integer}`
   = note: required for the cast to the object type `dyn futures_async_stream::__private::stream::Stream<Item = (i32, i32)> + Send`
   = note: this error originates in the attribute macro `stream` (in Nightly builds, run with -Z macro-backtrace for more info)
